Add plugin: Simple mind map

wanglin2 opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

Link to my plugin: https://github.com/wanglin2/obsidian-simplemindmap

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [x]  macOS
  - [ ]  Linux
  - [x]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

Zachatoo commented:
You reference several functions (`walk`, `nodeRichTextToTextWithWrap`, etc) that are not defined in your codebase. Are not all of your source files included in your repo? Please include all source files in your repo. I will continue my review after it's clear that all source files are included in your repo.

[# SimpleMindMap Obsidian Plugin](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/README_en.md#L3), [# SimpleMindMap Obsidian 插件](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/README.md#L3)
Naming something "Obsidian xyz" or "xyz Obsidian" is reserved for first party products we create, please change this header. You can do "xyz for Obsidian" or not include Obsidian in the name.

["fundingUrl": "",](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/manifest.json#L9)
This is meant for links to services like _Buy me a coffee_, _GitHub sponsors_ and so on, if you don't have such a link remove this line.

[{](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/locales/en.json#L1-L142)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

["minAppVersion": "0.15.0",](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/manifest.json#L5)
You are using newer API's, set this to the latest public build number.

[const scope = this.app.keymap.getRootScope()](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/Commands.js#L42-L58)
Please register these hotkeys to the scope of the relevant component, not to the root scope. Your components can unregister these hotkeys automatically on component unload.

[const dir = vault.getAbstractFileByPath(dirPath)](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/main.js#L422-L426), [const existingFile = vault.getAbstractFileByPath(filePath)](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/main.js#L112-L124), [const file = this.app.vault.getAbstractFileByPath(fileName)](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/Commands.js#L95-L105)
When using [`Vault.getAbstractFileByPath`](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getAbstractFileByPath), the result will be either a `TFile`, `TFolder`, or `null`. Your current code does not account for all of these scenarios. If you are only expecting a `TFile`, prefer to use [`Vault.getFileByPath`](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getFileByPath). If you are only expecting a `TFolder`, prefer to use [`Vault.getFolderByPath`](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getFolderByPath).

[hotkeys: [](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/Commands.js#L137-L142)
You should not provide a default hotkey. The reason being that it's easy to pick a hotkey that a user already has configured and they could get confused when that key doesn't do what they expect. Also, it's hard choosing a safe default hotkey that's available for all operating systems & does not conflict with existing plugins.

---

[if (typeof this.app.vault.setConfigFileIcon === 'function') {](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/main.js#L362-L364)
I'm not seeing this internal API in Obsidian 1.9.10. I'm not sure what version you're seeing this on, but I wouldn't recommend using this API.

[if (line.trim() === '---') {](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/metadataAndMarkdown.js#L21-L24)
This is going to return false positives, for example, a horizontal line somewhere after the frontmatter.

[if (currentFile?.extension !== 'md') {](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/Commands.js#L97-L101)
Should you check before creating the file that you'll be able to insert the link? If so, you should use an [editorCheckCallback](https://docs.obsidian.md/Plugins/User+interface/Commands#Editor+commands) and you should check that the editor is for a markdown file. The `ctx` provided by the callback will tell you if it's for a `MarkdownView` or something else.

[const markdownLeafs = this.app.workspace.getLeavesOfType(](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/Menus.js#L106-L115)
As of Obsidian v1.7.2, when Obsidian loads, all views are created as instances of **DeferredView**. Once a view is visible on screen (i.e. the tab is selected within its containing tab group), the `leaf` will re-render and the view will be switched out to the correct `View` instance. Please see [understanding deferred views](https://docs.obsidian.md/Plugins/Guides/Understanding+deferred+views) for how to safely access your custom view.

[const ref = this.app.metadataCache.on('changed', () => {](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/ob/Menus.js#L119-L120)
This could be triggered by a different file. Please validate that it's the correct file first, the [callback](https://docs.obsidian.md/Reference/TypeScript+API/MetadataCache/on('changed')) includes which file's cache was changed.

["axios": "^1.7.9",](https://github.com/wanglin2/obsidian-simplemindmap/blob/c70267660574b51209b18ff27a0b38a1d0c9ae1c/plugin/package.json#L43)
You don't need this nor are you using it, please remove this dependency.

Zachatoo commented:
I could've sworn there were some places that were importing those functions from `../utils`, not your other project, but I'm not seeing them. I'll continue reviewing this later.

Zachatoo commented:
I started reviewing this again and I'm seeing several changes I requested that you have not made yet, can you please finish making the changes I requested, or if you aren't going to make them, please explain why?

wanglin2 commented:
> I started reviewing this again and I'm seeing several changes I requested that you have not made yet, can you please finish making the changes I requested, or if you aren't going to make them, please explain why?

Hello, I just submitted and made the necessary modifications again. The following picture shows my current modification status:

<img width="837" height="669" alt="image" src="https://github.com/user-attachments/assets/0b125120-27ee-4155-9aa9-2aa8084d0253" />

<img width="830" height="490" alt="image" src="https://github.com/user-attachments/assets/f575bcbf-0241-4306-bde9-19eced979dc3" />

For some of the unmodified issues, let me explain:

1.hotkeys: Because the shortcut keys in my other project are all built-in in the library, I haven't made any significant changes yet

2.if (line.trim() === '---') {: This is because the content structure of the smm file is generated by the plugin and cannot be modified by the user (there are significant prompts in the file), so generally speaking, this situation does not occur

And currently, there are probably hundreds of people testing my plugin internally, and no one has reported these two issues yet

Zachatoo commented:
1. This is a requirement for submission, we've seen too many conflicting hotkeys with plugins in the past to allow plugins to set their own hotkeys anymore.
2. That's fine.

wanglin2 commented:
> 1. This is a requirement for submission, we've seen too many conflicting hotkeys with plugins in the past to allow plugins to set their own hotkeys anymore.
> 2. That's fine.

Hello, I have completed the modification of the shortcut keys, removed the default shortcut keys, and converted them into commands, allowing users to set their own shortcut keys. Could you please review it again when you have time? Thank you~

Zachatoo commented:
[this.contentEl.style.padding = 0](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/SmmEditView.js#L32), [emptyDiv.style.color = 'var(--text-muted)'](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/MarkdownPostProcessor.js#L232-L236), [style="width: 80px"](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/AssociativeLineStyle.vue#L33), and others, search for `style`
You should avoid assigning styles via JavaScript or in HTML and instead move all these styles into CSS so that they are more easily adaptable by themes and snippets.

[await this.plugin.app.vault.process(file, () => {](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/MarkdownPostProcessor.js#L180-L182)
[`Vault.process`](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+%60Vault.process%60+instead+of+%60Vault.modify%60+to+modify+a+file+in+the+background) should not be used in this way. Please refer to [this YouTube video](https://youtu.be/RvtV9DKr3ag?t=52) for how the process API works and how to properly use it.

[await this.app.vault.modify(renamedFile, mdStr)](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/Menus.js#L125)
Prefer to use [`Vault.process`](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+%60Vault.process%60+instead+of+%60Vault.modify%60+to+modify+a+file+in+the+background) to modify a file in the background.

[el.innerHTML = icon](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/SmmEditView.js#L750), [this.titleEl.innerHTML = newName](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/ModifyFileName.js#L34), [this.titleEl.innerHTML = originalName](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/ModifyFileName.js#L59), [linkEl.innerHTML = \`](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/SmmSettingTab.js#L766), [contentContainer.innerHTML = this.params.html](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/TextInfoDialog.js#L20), [return createFragment(frag => (frag.createDiv().innerHTML = html))](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/ob/utils.js#L126), [this.statusEl.innerHTML = ''](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/Count.vue#L37), [countEl.innerHTML = this.textStr](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/Count.vue#L48), [this.statusEl.innerHTML = \`](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/Count.vue#L52), [el.innerHTML = text](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/Edit.vue#L452), [e.target.innerHTML = beforeText](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/Outline.vue#L230), [e.target.innerHTML = beforeText](https://github.com/wanglin2/obsidian-simplemindmap/blob/6bc3b85c8ddcfe4da77bdc0a84f74cfd87910267/plugin/src/pages/Edit/components/OutlineEdit.vue#L228)
[Avoid using `innerHTML`](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%60innerHTML%60%2C+%60outerHTML%60+and+%60insertAdjacentHTML%60)
